aten/src/ATen/ExpandUtils.h

```diff
@@ -103,46 +103,128 @@ inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>> expand_inpla
 inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>> expand_inplace(const Tensor &tensor, const Tensor &to_expand1, Tensor &&to_expand2, const char *api_name) = delete;
 inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>> expand_inplace(const Tensor &tensor, Tensor &&to_expand1, Tensor &&to_expand2, const char *api_name) = delete;
 
-
-inline std::tuple<Tensor, Tensor> expand_outplace(const Tensor &to_expand1, const Tensor &to_expand2) {
+// See NOTE [ ExpandUtils Borrowing ] above for `MaybeOwned` explanation.
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(const Tensor &to_expand1, const Tensor &to_expand2) {
   if (to_expand1.sizes().equals(to_expand2.sizes())) {
-    return std::make_tuple(to_expand1, to_expand2);
+    return std::make_tuple(
+        c10::MaybeOwned<Tensor>::borrowed(to_expand1),
+        c10::MaybeOwned<Tensor>::borrowed(to_expand2));
   }
 
   auto expanded_size = infer_size_dimvector(to_expand1.sizes(), to_expand2.sizes());
   return std::make_tuple(
-      to_expand1.expand(expanded_size, /*implicit=*/true), // see [expand implicit]
-      to_expand2.expand(expanded_size, /*implicit=*/true));
+      c10::MaybeOwned<Tensor>::owned(to_expand1.expand(expanded_size, /*implicit=*/true)), // see [expand implicit]
+      c10::MaybeOwned<Tensor>::owned(to_expand2.expand(expanded_size, /*implicit=*/true)));
 }
 
-inline std::tuple<Tensor, Tensor> expand_outplace(const Tensor &to_expand1, const Tensor &to_expand2, const char *api_name) {
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(Tensor &&to_expand1, const Tensor &to_expand2) = delete;
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(const Tensor &to_expand1, Tensor &&to_expand2) = delete;
+inline std::tuple<c10::MaybeOwned<Tensor>, c10::MaybeOwned<Tensor>>
+expand_outplace(Tensor &&to_expand1, Tensor &&to_expand2) = delete;
```

Comment:
I wonder if there is anything we could do with implicit conversions to avoid having to explicitly delete all combinatorial combinations of `Tensor&&` overloads, cuz it kind of bytes having a two argument function balloon into four variants here.

Reply:
And it's even worse for three argument functions

Reply:
Implicit conversions are not quite good enough. I tried adding the following helper:

```
// NonRvalueArgument<T> can be used instead of const T& as an argument
// type if you wish to disallow rvalues (e.g., a prvalue temporary)
// from being passed. The advantage is that you avoid having to
// explicitly delete 2**N - 1 overloads given N such arguments.
template <typename T>
class NonRvalueArgument {
  const T& ref;
 public:
  /* implicit */ NonRvalueArgument(const T& r) : ref(r) {}
  NonRvalueArgument(T&& r) = delete;

  operator const T&() const {
    return ref;
  }
};
```

But you cannot call `Tensor` methods on a `NonRvalueArgument<Tensor>`. We would have to implement `operator*` and `operator->` and use pointer syntax. If you would prefer that, I can send out a diff to do it, but either way there is something ugly involved, and with the deleted overloads they're at least not cluttering the function body.

Reply:
Well, at least pointer syntax doesn't require 2**N overloads. NBD, you can land this as is, but if you ever hit a 4-ary probably crack this one open.